staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/clock/testing/fake.go

```diff
@@ -211,9 +175,22 @@ func (fec *FakeEventClock) SetTime(t time.Time) {
 	}
 }
 
+// Sleep returns after the given duration has passed.
+// Unlike the base FakeClock's Sleep, this method does not itself advance the clock
+// but rather leaves that up to other actors (e.g., Run).
+func (fec *FakeEventClock) Sleep(duration time.Duration) {
+	doneCh := make(chan struct{})
+	fec.EventAfterDuration(func(time.Time) {
+		fec.clientWG.Add(1)
+		close(doneCh)
+	}, duration)
+	fec.clientWG.Add(-1)
```

Comment:
Should be vice-versa, right?

We should first increase by 1 and in the event handler substract 1 [otherwise we can in theory go below 0, which is forbidden].

Also - the addition should go before the EventAfterDuration.

Reply:
Interesting.  I see now that this is not a general Sleep, but only one that can be invoked from a goroutine whose activity has been counted.  That's why the `Add(-1)` will not take the count below zero.  The `Add(1)` is right because it is counting the re-activation of that goroutine.

Reply:
Updated comments on FakeEventClock and FakeEventClock::Sleep to clearly state that restriction, and also to more clearly document the association between the FakeEventClock and its GoRoutineCounter.

Reply:
I also added commentary on the `FakeEventClock::EventAfterDuration` and `FakeEventClock::EventAfterTime` methods to explicitly delineate counting responsibilities.